prow/test/integration/test/setup.go

```diff
@@ -0,0 +1,99 @@
+// +build e2etest
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"flag"
+	"fmt"
+	"os"
+	"os/user"
+	"path"
+	"strings"
+	"testing"
+
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	defaultNamespace = "default"
+	testpodNamespace = "test-pods"
+)
+
+var clusterContext = flag.String("cluster", "kind-kind-prow-integration", "The context of cluster to use for test")
+
+func getClusterContext() string {
+	return *clusterContext
+}
+
+func getDefaultKubeconfig(cfg string) string {
+	if cfg := strings.TrimSpace(cfg); cfg != "" {
+		return cfg
+	}
+	defaultKubeconfig := os.Getenv("KUBECONFIG")
+
+	// If KUBECONFIG env var isn't set then look for $HOME/.kube/config
+	if defaultKubeconfig == "" {
+		if usr, err := user.Current(); err == nil {
+			defaultKubeconfig = path.Join(usr.HomeDir, ".kube/config")
+		}
+	}
+	return defaultKubeconfig
+}
+
+func NewClients(configPath, clusterName string) (ctrlruntimeclient.Client, error) {
+	cfg, err := BuildClientConfig(getDefaultKubeconfig(configPath), clusterName)
+	if err != nil {
+		return nil, fmt.Errorf("failed create rest config: %v", err)
+	}
+	return ctrlruntimeclient.New(cfg, ctrlruntimeclient.Options{})
+}
+
+func NewClientsFromConfig(cfg *rest.Config) (*kubernetes.Clientset, error) {
+	kubeClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return kubeClient, nil
+}
+
+func BuildClientConfig(kubeConfigPath, clusterName string) (*rest.Config, error) {
+	overrides := clientcmd.ConfigOverrides{}
+	// Override the cluster name if provided.
+	if clusterName != "" {
+		overrides.Context.Cluster = clusterName
```

Comment:
Agreed that `clusterName` is actually a context, this just happens to also be named as clustername in `clientcmd.ConfigOverrides`. The documentation didn't clearly mention which one to use, but the code examples I have seen somewhere else uses `Context.Cluster`, I have stepped in debugging mode and confirmed that it worked.

Reply:
Doesn't this only work because they happen to be the same and the current-context is already set? I would expect this to work without this override. Overriding just the cluster name without overriding the user probably wouldn't work if it was actually making a change. Here are the godoc: https://godoc.org/k8s.io/client-go/tools/clientcmd#ConfigOverrides

Reply:
I see your point, also override currentcontext

Reply:
I'm pretty sure overwriting the `Context.Cluster` would be problematic if the values actually differed since the cluster needs to be associated with the correct user (`AuthInfo`). That being said I don't know if the values will ever differ in practice so this might be fine anyways.